### PR TITLE
fix: drop bounding caps by default if unset

### LINKED ISSUE #3434: [Bug]: failed to drop capabilities in youki exec
### Bug Description

Youki exec fails with "failed to drop capabilities" where `runc run` (not `exec`) fails.

### Steps to Reproduce

`config.json` lists all possible capabilities:

```
{
  "ociVersion": "1.2.0",
  "process": {
    "terminal": false,
    "user": {
      "uid": 0,
      "gid": 0,
      "umask": 18
    },
    "args": [
      "/bin/sh"
    ],
    "env": [
      "DEBIAN_FRONTEND=noninteractive",
      "HOSTNAME=hds-c2ftv93mtzsz",
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    ],
    "cwd": "/",
    "capabilities": {
      "effective": [
        "CAP_CHOWN",
        "CAP_DAC_OVERRIDE",
        "CAP_DAC_READ_SEARCH",
        "CAP_FOWNER",
        "CAP_FSETID",
        "CAP_KILL",
        "CAP_SETGID",
        "CAP_SETUID",
        "CAP_SETPCAP",
        "CAP_LINUX_IMMUTABLE",
        "CAP_NET_BIND_SERVICE",
        "CAP_NET_BROADCAST",
        "CAP_NET_ADMIN",
        "CAP_NET_RAW",
        "CAP_IPC_LOCK",
        "CAP_IPC_OWNER",
        "CAP_SYS_MODULE",
        "CAP_SYS_RAWIO",
        "CAP_SYS_CHROOT",
        "CAP_SYS_PTRACE",
        "CAP_SYS_PACCT",
        "CAP_SYS_ADMIN",
        "CAP_SYS_BOOT",
        "CAP_SYS_NICE",
        "CAP_SYS_RESOURCE",
        "CAP_SYS_TIME",
        "CAP_SYS_TTY_CONFIG",
        "CAP_MKNOD",
        "CAP_LEASE",
        "CAP_AUDIT_WRITE",
        "CAP_AUDIT_CONTROL",
        "CAP_SETFCAP",
        "CAP_MAC_OVERRIDE",
        "CAP_MAC_ADMIN",
        "CAP_SYSLOG",
        "CAP_WAKE_ALARM",
        "CAP_BLOCK_SUSPEND",
        "CAP_AUDIT_READ",
        "CAP_PERFMON",
        "CAP_BPF",
        "CAP_CHECKPOINT_RESTORE"
      ],
      "inheritable": [
        "CAP_CHOWN",
        "CAP_DAC_OVERRIDE",
        "CAP_DAC_READ_SEARCH",
        "CAP_FOWNER",
        "CAP_FSETID",
        "CAP_KILL",
        "CAP_SETGID",
        "CAP_SETUID",
        "CAP_SETPCAP",
        "CAP_LINUX_IMMUTABLE",
        "CAP_NET_BIND_SERVICE",
        "CAP_NET_BROADCAST",
        "CAP_NET_ADMIN",
        "CAP_NET_RAW",
        "CAP_IPC_LOCK",
        "CAP_IPC_OWNER",
        "CAP_SYS_MODULE",
        "CAP_SYS_RAWIO",
        "CAP_SYS_CHROOT",
        "CAP_SYS_PTRACE",
        "CAP_SYS_PACCT",
        "CAP_SYS_ADMIN",
        "CAP_SYS_BOOT",
        "CAP_SYS_NICE",
        "CAP_SYS_RESOURCE",
        "CAP_SYS_TIME",
        "CAP_SYS_TTY_CONFIG",
        "CAP_MKNOD",
        "CAP_LEASE",
        "CAP_AUDIT_WRITE",
        "CAP_AUDIT_CONTROL",
        "CAP_SETFCAP",
        "CAP_MAC_OVERRIDE",
        "CAP_MAC_ADMIN",
        "CAP_SYSLOG",
        "CAP_WAKE_ALARM",
        "CAP_BLOCK_SUSPEND",
        "CAP_AUDIT_READ",
        "CAP_PERFMON",
        "CAP_BPF",
        "CAP_CHECKPOINT_RESTORE"
      ],
      "permitted": [
        "CAP_CHOWN",
        "CAP_DAC_OVERRIDE",
        "CAP_DAC_READ_SEARCH",
        "CAP_FOWNER",
        "CAP_FSETID",
        "CAP_KILL",
        "CAP_SETGID",
        "CAP_SETUID",
        "CAP_SETPCAP",
        "CAP_LINUX_IMMUTABLE",
        "CAP_NET_BIND_SERVICE",
        "CAP_NET_BROADCAST",
        "CAP_NET_ADMIN",
        "CAP_NET_RAW",
        "CAP_IPC_LOCK",
        "CAP_IPC_OWNER",
        "CAP_SYS_MODULE",
        "CAP_SYS_RAWIO",
        "CAP_SYS_CHROOT",
        "CAP_SYS_PTRACE",
        "CAP_SYS_PACCT",
        "CAP_SYS_ADMIN",
        "CAP_SYS_BOOT",
        "CAP_SYS_NICE",
        "CAP_SYS_RESOURCE",
        "CAP_SYS_TIME",
        "CAP_SYS_TTY_CONFIG",
        "CAP_MKNOD",
        "CAP_LEASE",
        "CAP_AUDIT_WRITE",
        "CAP_AUDIT_CONTROL",
        "CAP_SETFCAP",
        "CAP_MAC_OVERRIDE",
        "CAP_MAC_ADMIN",
        "CAP_SYSLOG",
        "CAP_WAKE_ALARM",
        "CAP_BLOCK_SUSPEND",
        "CAP_AUDIT_READ",
        "CAP_PERFMON",
        "CAP_BPF",
        "CAP_CHECKPOINT_RESTORE"
      ]
    }
  },
  "root": {
    "path": "rootfs"
  },
  "hostname": null,
  "mounts": [
    {
      "destination": "/proc",
      "type": "proc",
      "source": "proc",
      "options": [
        "nosuid",
        "noexec",
        "nodev"
      ]
    },
    {
      "destination": "/dev",
      "type": "tmpfs",
      "source": "tmpfs",
      "options": [
        "nosuid",
        "strictatime",
        "mode=755",
        "size=65536k"
      ]
    },
    {
      "destination": "/sys",
      "type": "sysfs",
      "source": "sysfs",
      "options": [
        "nosuid",
        "noexec",
        "nodev",
        "ro"
      ]
    },
    {
      "destination": "/dev/pts",
      "type": "devpts",
      "source": "devpts",
      "options": [
        "nosuid",
        "noexec",
        "newinstance",
        "ptmxmode=0666",
        "mode=0620",
        "gid=5"
      ]
    },
    {
      "destination": "/dev/shm",
      "type": "tmpfs",
      "source": "shm",
      "options": [
        "nosuid",
        "noexec",
        "nodev",
        "mode=1777",
        "size=67108864"
      ]
    },
    {
      "destination": "/dev/mqueue",
      "type": "mqueue",
      "source": "mqueue",
      "options": [
        "nosuid",
        "noexec",
        "nodev"
      ]
    },
    {
      "destination": "/sys/fs/cgroup",
      "type": "cgroup",
      "source": "cgroup",
      "options": [
        "rprivate",
        "nosuid",
        "noexec",
        "nodev",
        "relatime",
        "rw"
      ]
    }
  ],
  "linux": {
    "namespaces": [
      {
        "type": "pid"
      },
      {
        "type": "ipc"
      },
      {
        "type": "uts"
      },
      {
        "type": "mount"
      },
      {
        "type": "cgroup"
      }
    ],
    "cgroupsPath": "/ch-inner"
  }
}
```

Not sure what is the right behavior here. Anyway,

```
$ youki run xx
```

starts fine, but then

```
$ youki exec xx pwd
ERROR libcontainer::process::init::process: failed to drop capabilities err=SetCaps(CapsError("capset failure: Operation not permitted (os error 1)"))
ERROR libcontainer::process::container_intermediate_process: failed to initialize container process: failed syscall
ERROR libcontainer::process::container_main_process: failed to wait for init ready: exec process failed with error error in executing process : failed syscall
ERROR libcontainer::container::builder_impl: failed to run container process exec process failed with error error in executing process : failed syscall
ERROR youki: error in executing command: failed to create container: exec process failed with error error in executing process : failed syscall
exec failed : failed to create container: exec process failed with error error in executing process : failed syscall
```

But

```
$ runc run xx
ERRO[0000] runc run failed: unable to start container process: error during container init: unable to apply caps: can't apply capabilities: operation not permitted
```

### Expectation

I presume that either
- container should not be able to start
- or if container started, then `exec` should be successful

### System and Setup Info

youki version: 0.5.7
commit: 0.5.7-bd54457ba9de1629f21fe678687f433aec63e7c7

runc version 1.3.0
commit: v1.3.0-0-g4ca628d1
spec: 1.2.1
go: go1.24.7
libseccomp: 2.5.5

6.14.0-1015-gcp #16-Ubuntu
(running on GCE VM)

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by YawKar**
## Description
This PR fixes #3434: youki's `run` and `exec` paths handled a missing `.process.capabilities.bounding` inconsistently.

The table below shows the current state of things:
| command | reaction to missing bounding caps | result |
|---------|-----------------------------------|--------|
| sudo runc run | explicitly drop all bounding, then capset fails | fails |
| sudo youki run | don't touch bounding | succeeds |
| sudo youki exec | explicitly drop all bounding, then capset fails | fails (at exec!) |

The chosen way to fix it is essentially the same as it is in runc ([libcontainer/capabilities/capabilities.go#L100-L107](https://github.com/opencontainers/runc/blob/d0aeb9e3e22e02eec9f9a42416267a4f358cc571/libcontainer/capabilities/capabilities.go#L100-L107), [called here: libcontainer/init_linux.go#L340-L343](https://github.com/opencontainers/runc/blob/d0aeb9e3e22e02eec9f9a42416267a4f358cc571/libcontainer/init_linux.go#L340-L343)). Always clear bounding caps and then set only what was given (or nothing in case bounding is unset), after that apply.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - it breaks all configs that have eff/inh/perm caps set but miss bounding (mostly hand-crafted oci configs)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [x] Test updates
- [ ] CI/CD related changes
- [ ] Other (please describe):

## Testing

- [ ] Added new unit tests
- [x] Added new integration tests
- [x] Ran existing test suite
- [ ] Tested manually (please provide steps)

## Related Issues

Fixes #3434

### CODE REVIEW

**Comment by YawKar** [comment]
@saku3 👋

This implements option 1 as you suggested. Could you take a look when you have a chance?

File: crates/libcontainer/src/capabilities.rs [hunk 1]
```diff
@@ -135,10 +135,10 @@ pub fn drop_privileges<S: Syscall + ?Sized>(
     cs: &LinuxCapabilities,
     syscall: &S,
 ) -> Result<(), SyscallError> {
-    if let Some(bounding) = cs.bounding() {
-        tracing::debug!("dropping bounding capabilities to {:?}", bounding);
-        syscall.set_capability(CapSet::Bounding, &to_set(bounding))?;
-    }
+    let empty = Default::default();
```
Comment (nayuta723) [inline crates/libcontainer/src/capabilities.rs]: [imo] It makes more sense to change the behavior in `oci-spec-rs` rather than fixing it on the `youki` side.
https://github.com/youki-dev/oci-spec-rs/blob/main/src/runtime/process.rs#L363-L383
Also, even if we were to fix this on the youki side, it would probably make more sense to add it to `https://github.com/youki-dev/youki/blob/main/crates/youki/src/commands/spec_json.rs`.

ref: #3560

**Review by nayuta723 — COMMENTED** [review]
Thank you for the contribution. I’ve left a comment with my thoughts. Please take a look.

File: crates/libcontainer/src/capabilities.rs (hunk 1 above)
Comment (YawKar) [reply crates/libcontainer/src/capabilities.rs]: Hmm, that's strange. I haven't found any defaults in actual OCI runtime spec regarding capabilities. 

Maybe these are for `youki spec`? For generation, not for "what to bring in if unset in config.json"?

runc drops all bounding capabilities when they're missing in config.json

Comment (YawKar) [reply crates/libcontainer/src/capabilities.rs]: What I want to say is there seems to be 2 defaults:
- default for generation of a new spec by `youki spec`: that can do whatever it wants
- default bounding capabilities for an already generated by something else config.json with unset bounding caps

I'm not sure but from my POV it looks like you try to highlight the former while I'm solving the latter here.

**Comment by YawKar** [comment]
@nayuta723 hi, thank you for your work on collecting differences between youki spec and runc spec!

But I believe this is not the case here. In my investigation I did indeed use `youki spec` but only as a mean to generate a minimal config.json, I then patched it manually to remove bounding caps to recreate the issue at hand.

File: crates/libcontainer/src/capabilities.rs (hunk 1 above)
Comment (nayuta723) [reply crates/libcontainer/src/capabilities.rs]: Sorry, I misunderstood. I’ll take another look.

Comment (nayuta723) [inline crates/libcontainer/src/capabilities.rs]:
```suggestion
    let empty_caps = Default::default();
```

File: crates/libcontainer/src/capabilities.rs [hunk 2]
```diff
@@ -135,10 +135,10 @@ pub fn drop_privileges<S: Syscall + ?Sized>(
     cs: &LinuxCapabilities,
```
Comment (nayuta723) [inline crates/libcontainer/src/capabilities.rs:135]: Please also add a test case to `test_drop_privileges`.

Comment (nayuta723) [reply crates/libcontainer/src/capabilities.rs:135]: It would also be nice to add a test case for when `bounding` is `None` in the `run` or `exec` tests, although that can be done in a separate PR.

**Review by nayuta723 — COMMENTED** [review]
Sorry for the confusion earlier. I’ve reviewed it now, so please take a look when you have a time.

File: crates/libcontainer/src/capabilities.rs (hunk 2 above)
Comment (YawKar) [reply crates/libcontainer/src/capabilities.rs:135]: @nayuta723 this PR has already introduced 2 integration tests in contest for `run` path, 1 positive and 1 negative. Could you please share with me more details on what do you mean by `run` tests?

As for `exec` path, agreed, it's not that hard so I'll just add `bounding: None` tests here: https://github.com/youki-dev/youki/blob/main/tests/contest/contest/src/tests/exec/cap_test.rs

Comment (YawKar) [reply crates/libcontainer/src/capabilities.rs:135]: @nayuta723 added test for when bounding is None (not empty) to `exec` integration tests that checks that bounding caps are indeed zeroed.

Comment (nayuta723) [reply crates/libcontainer/src/capabilities.rs:135]: > this PR has already introduced 2 integration tests in contest for run path, 1 positive and 1 negative. Could you please share with me more details on what do you mean by run tests?

Indeed, `test_inside_container` is effectively testing the same behavior as `run`. I also checked, and it doesn't seem like there are plans to add dedicated tests for `run`, so I think we're good. Thanks!

**Comment by YawKar** [comment]
@saku3 ~~hmm, I don't know why but CI is stuck :(~~
~~Do you know what status it is waiting for? The only thing I see missing from the workflow is "changes" which aren't listed in the checks list.~~

~~May I ask you to force-launch them, please?~~

I'll try to rebase onto main.
UPD: it's unstuck now

**Review by nayuta723 — APPROVED** [review]
LGTM! Thanks!

@saku3 
Please take a look when you have some time!

**Review by saku3 — APPROVED** [review]
